### PR TITLE
Add in small optimization for instr comparison

### LINKED ISSUE #10578: [FEA] Support project expression rewrite for the case ```stringinstr(str_col, substr) > 0``` to ```contains(str_col, substr)```
**Is your feature request related to a problem? Please describe.**
In the case of ```stringinstr(str, substr) > 0```, currently it's interpreted as ```GreaterThan``` with ```stringinstr``` as its child expression.  And in this case, we don't care about the string location, thus, we can rewrite this condition with ```contains``` at the physical plan level within Spark Rapids Plugin, avoiding an extra binary operation. And ```binary_operation``` is more expensive than string contains.

### PR DESCRIPTION
**Pull request by revans2**
In older versions of Spark contains was not exposed, so we still see some people use tricks like `instr > 0`. This optimizes that operations. It is not a huge win. It just removes a > comparison that is not needed.

This fixes #10578

### CODE REVIEW

**Comment by revans2** [comment]
build

**Comment by revans2** [comment]
I should say that I ran some performance numbers, and like you would expect, it really comes down to the amount of time instr is processing the data. So for small strings I saw a 1.6x speedup. For very large strings it went down to just a 1.1x speedup.  Note that this is end to end for a single query that did 20 instr comparisons that could all be converted into contains.

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 1]
```diff
@@ -1962,6 +1962,29 @@ case class GpuStringToMap(strExpr: Expression,
...
 }
 
+object GpuStringInstr {
+  def optimizeContains(cmp: GpuExpression): GpuExpression = {
+    cmp match {
+      case GpuGreaterThan(GpuStringInstr(str, substr: GpuLiteral), GpuLiteral(v, _)) if v == 0 =>
+        // stringInstr(A, B) > 0 becomes contains(A, B)
+        GpuContains(str, substr)
+      case GpuGreaterThanOrEqual(GpuLiteral(v, _), GpuStringInstr(str, substr: GpuLiteral))
+        if v == 0 =>
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:1972]:
```suggestion
        if v == 1 =>
```

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala [hunk 2]
```diff
@@ -1962,6 +1962,29 @@ case class GpuStringToMap(strExpr: Expression,
   }
 }
 
+object GpuStringInstr {
+  def optimizeContains(cmp: GpuExpression): GpuExpression = {
+    cmp match {
+      case GpuGreaterThan(GpuStringInstr(str, substr: GpuLiteral), GpuLiteral(v, _)) if v == 0 =>
```
Comment (gerashegalov) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:1968]: nit: here and below we could put value into the pattern, I think it is readable with your comments 
```suggestion
      case GpuGreaterThan(GpuStringInstr(str, substr: GpuLiteral), GpuLiteral(0, _)) =>
```

**Comment by winningsix** [comment]
> cessing the data. So for small strings I saw a 1.6x speedup. For very large strings it went down to just a 1.1x speedup. Note that this is end to end for a single query that did 20 instr comparisons that could all be converted into contains.

Just curious. Most benefits come from the elimination of binary operation "greatThan", right?

**Comment by revans2** [comment]
> Just curious. Most benefits come from the elimination of binary operation "greatThan", right?

Yes, that is what I saw. It shifted the time from one kernel to another. We are also going to investigate if there are things we can do to speed up these kernels, especially for long strings.

**Comment by revans2** [comment]
@gerashegalov @winningsix @firestarman I have addressed the review comments. Thanks for picking up on my mistakes. I have fixed them and manually verified that we are doing the right thing in all cases.

**Comment by revans2** [comment]
build

**Review by gerashegalov — APPROVED** [review]
LGTM

**Review by firestarman — APPROVED** [review]

**Review by winningsix — APPROVED** [review]

**Pull request merged**